### PR TITLE
syntax: Make `_` a reserved identifier

### PR DESCRIPTION
**Pull request by petrochenkov**
Why:
- Lexically `_` is an identifier.
- Internally it makes implementation of `use Trait as _;` (https://github.com/rust-lang/rust/issues/48216) and some other things cleaner.
- We prevent the externally observable effect of `_` being accepted by macros expecting `ident` by treating `_` specially in the `ident` matcher:
```rust
macro_rules! m {
    ($i: ident) => { let $i = 10; }
}

m!(_); // Still an error
```

### CODE REVIEW

**Comment by petrochenkov** [comment]
cc @rust-lang/lang 
r? @nikomatsakis

File: src/libsyntax_pos/symbol.rs [hunk 1]
```diff
@@ -261,73 +261,77 @@ macro_rules! declare_keywords {(
 declare_keywords! {
     // Special reserved identifiers used internally for elided lifetimes,
     // unnamed method parameters, crate root module, error recovery etc.
-    (0,  Invalid,        "")
-    (1,  CrateRoot,      "{{root}}")
-    (2,  DollarCrate,    "$crate")
+    (0,  Invalid,            "")
+    (1,  CrateRoot,          "{{root}}")
+    (2,  DollarCrate,        "$crate")
+    (3,  Underscore,         "_")
```
Comment (petrochenkov) [inline src/libsyntax_pos/symbol.rs]: I can move it into the keyword list below, the only difference would be whether error messages say "keyword `_`" or "reserved identifier `_`".

**Comment by rkruppe** [comment]
`_` is different enough from identifiers that I think it's pedagogically harmful to treat it as one in any contexts. One example where I've seen this come up is in patterns &mdash; `let _ = <value with drop glue>` dropping immediately rather than at the end of the block, `&Foo(_)` not causing a "cannot move out of borrowed content" error, etc.

File: src/libsyntax_pos/symbol.rs (hunk 1 above)
Comment (Mark-Simulacrum) [reply src/libsyntax_pos/symbol.rs]: I think "reserved identifier" is better than "keyword" as it doesn't really look/feel like a keyword.

**Comment by nikomatsakis** [comment]
I agree with @rkruppe. I don't want `_` to be considered an identifer from the user's perspective -- I've although of it as a special token that has special behavior in places where it appears.

**Comment by Centril** [comment]
I also think this also conflicts with one of the alternatives to `PhantomData<T>` proposed here: https://github.com/rust-lang/rfcs/pull/2357#issuecomment-371117332.

**Comment by petrochenkov** [comment]
@Centril 
It's still a *reserved* identifier (a keyword, basically), so it can't be used as a general purpose identifier and in the future we can interpret it as we want in any position where it doesn't have existing meaning.

`_` being an identifier and not a punctuation token should be an implementation detail except for the interaction with `ident` matcher (maybe it's also visible for procedural macros?).
Actually, it can be make it 100% implementation detail by tweaking `ident` matcher logic to keep the old behavior and not accept `_`.

**Comment by nikomatsakis** [comment]
@petrochenkov 

> Actually, it can be make it 100% implementation detail by tweaking ident matcher logic to keep the old behavior and not accept _.

👍 to that plan, this is what I was going to suggest.

**Comment by petrochenkov** [comment]
Updated with `_` being rejected by `ident` matcher in macros

**Comment by petrochenkov** [comment]
>[01:03:40] failures:
>[01:03:40]     [run-pass] run-pass/macro-first-set.rs

Ok, I guess I'll have to tweak the `may_begin_with` logic as well.

File: src/libsyntax/ext/tt/macro_parser.rs [hunk 2]
```diff
@@ -725,7 +725,7 @@ fn may_begin_with(name: &str, token: &Token) -> bool {
     match name {
         "expr" => token.can_begin_expr(),
         "ty" => token.can_begin_type(),
-        "ident" => token.is_ident(),
+        "ident" => token.is_ident() && !token.is_keyword(keywords::Underscore),
```
Comment (nikomatsakis) [inline src/libsyntax/ext/tt/macro_parser.rs]: can we add some sort of helper for this check, like `is_macro_ident`?

File: src/libsyntax/ext/tt/macro_parser.rs [hunk 3]
```diff
@@ -815,8 +814,9 @@ fn parse_nt<'a>(p: &mut Parser<'a>, sp: Span, name: &str) -> Nonterminal {
         "expr" => token::NtExpr(panictry!(p.parse_expr())),
         "ty" => token::NtTy(panictry!(p.parse_ty())),
         // this could be handled like a token, since it is one
+        // we prohibit passing `_` to macros expecting `ident` for now
         "ident" => match p.token {
-            token::Ident(sn) => {
+            token::Ident(sn) if sn.name != keywords::Underscore.name() => {
```
Comment (nikomatsakis) [inline src/libsyntax/ext/tt/macro_parser.rs]: we could then use the helper here too, rather than open-coding this

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Well, the code seems fine modulo one nit -- but remind me, what is the motivation for this change again? =)

**Comment by nikomatsakis** [comment]
Wait: one other question .. is `'_` also an ident? If so, can macro idents match against it?

**Comment by nikomatsakis** [comment]
Re-read motivation from the intro. r=me with nit addressed -- presuming `'_` is not being matched by `$x:ident`.

**Comment by petrochenkov** [comment]
@nikomatsakis 
`'_` is matched by `$lt: lifetime` actually (even without this patch).

We can prohibit this for consistency (both `'_` and `lifetime` matcher are unstable), but I'd rather eventually accept both `'_`/`_` as `lifetime`/`ident` instead.

**Comment by petrochenkov** [comment]
>what is the motivation for this change again

One more important thing that I didn't mention - `_` in `use Trait as _;` actually needs hygiene data attached to it!
With `_` being an identifier it happens automatically, but with `_` being a non-identifier token we'd have to do some special tracking.

**Comment by eddyb** [comment]
@petrochenkov @jseyfried IMO we should just switch to relying on `Span` for hygiene everywhere.

**Comment by petrochenkov** [comment]
@eddyb 
>we should just switch to relying on Span for hygiene everywhere.

This reminded me of something - `Span`s are now `u32`.
So we can change `Ident` representation from `{ Symbol, SyntaxContext }` to `{ Symbol, Span }` thus giving all identifiers spans (this is really great for diagnostics) without increasing the size of `Ident`.
This would also eliminate `SpannedIdent` as duplicated data and would allow to meaningfully reintroduce `Ident` into HIR thus eliminating some [horrible side tables](https://github.com/rust-lang/rust/blob/d089fe974edc53fa34384e8e76eeb1eca0d89042/src/libsyntax_pos/hygiene.rs#L463-L478).
Oh man, this is going to be great, I think I know what I'm going to spend this Saturday on.

(This is totally orthogonal to this PR though, it would still be more convenient for `_` to be an `Ident` even with that scheme.)

**Comment by petrochenkov** [comment]
Question: given that https://github.com/rust-lang/rust/pull/48942 is already in flight, should `r#_` be supported as a way to create a normal identifier named "_" or not?

**Comment by nikomatsakis** [comment]
@petrochenkov 

> We can prohibit this for consistency

I believe that indeed `'_` should *not* match `$lt:lifetime`. It is not a lifetime (in my mind) -- rather, it is a special token that is used to control the elision rules.

**Comment by nikomatsakis** [comment]
Well, I'm mildly reconsidering that -- see [this comment](https://github.com/rust-lang/rust/issues/34303#issuecomment-373439161)

**Pull request merged**